src/wp-includes/default-filters.php

```diff
@@ -521,7 +521,7 @@
  */
 // Theme.
 add_action( 'setup_theme', 'create_initial_theme_features', 0 );
-add_action( 'setup_theme', '_add_default_theme_supports', 1 );
+add_action( 'after_setup_theme', '_add_default_theme_supports', 1 );
```

Comment:
From what I remember (been a while), this hook was selected to align to the above hook that creates the theme features. Notice the priority order:

* 0 for creating the initial theme features
* 1 for adding the default theme supports

Here's the PR that added it https://github.com/WordPress/wordpress-develop/pull/2029.

Reply:
@hellofromtonya I understand that the `_add_default_theme_supports()` function must be called after `create_initial_theme_features()`, but I don't think it needs to be _right_ after.

Looking at the PR #2029, I don't see any discussion or reasoning on _why_ it was decided to use the `setup_theme` action immediately after the `create_initial_theme_features()` callback. Given the lack of an explanation, I assume it was simply chosen because that's what `create_initial_theme_features()` used, and because the `_add_default_theme_supports()` should fire after that.

The problem with that approach is: While the available theme features in general should be _registered_ on `setup_theme`, actually _adding_ theme supports based on the current theme needs to happen only on `after_setup_theme`, as outlined in the PR description.

Reply:
>While the available theme features in general should be registered on setup_theme, actually adding theme supports based on the current theme needs to happen only on `after_setup_theme`, as outlined in the PR description.

Why should Core itself register to `'after_theme_setup'`? Shouldn't it set itself up before themes get loaded into memory?

Because without the activated theme loaded into memory, how would Core know what default supports to add? `_add_default_theme_supports()` checks if the activated theme is a block theme. Aha.

Reply:
> Why should Core itself register to `'after_theme_setup'`? Shouldn't it set itself up before themes get loaded into memory?

Because before the `after_setup_theme`, the current theme is still being set up, including potentially overriding what the current theme is, as the Customizer does. So only after the `setup_theme` action (i.e. during the `after_setup_theme` action) we truly know what the current theme is.

It's also worth noting that all features that `_add_default_theme_supports()` adds are not block theme specific. These features are added in classic themes as well, including all classic default themes of the past 5+ years. The themes do it on `after_setup_theme`, so core should do it too. The reason that core even handles it for block themes is to avoid the need for block theme developers to implement a `functions.php` file just for these default features, which improves the developer experience.

Reply:
>It's also worth noting that all features that _add_default_theme_supports() adds are not block theme specific.

For clarity, I was referring to:

```php
function _add_default_theme_supports() {
	if ( ! wp_is_block_theme() ) {
		return;
	}
```

where the function bails out if the theme is not a block theme. More specifically, I was mentally wrestling with: what is the state of that check when hooked into `setup_theme`.

Was it working as intended?

Backtracing the logic, yes, it does work, because `wp_get_theme()` can be populated by the activated theme in the `'stylesheet'` option. Thus, the theme did not need to be set up for this to work.

That part of the logic working though is not necessarily relevant to the bug being reported.

Reply:
Let me clarify with an example of how this being hooked into `setup_theme` does not work as expected:
1. Your site uses a block theme (e.g. Twenty Twenty-Three).
2. On _Appearance > Themes_, you click "Live Preview" for a classic theme, e.g. Twenty Twenty-One.
3. In the live preview, Twenty Twenty-One will now have all the "theme supports" from `_add_default_theme_supports()`, which is a bug, as those theme supports should only be added when the theme is a block theme.

The reason for that is that at the `setup_theme` action, WordPress still "thinks" the current theme is a block theme and thus adds the theme supports. The Customizer however overrides the current theme for the preview to be the classic theme only after that, and this logic has been present in WordPress core for many many years, far longer than when this bug was introduced. Therefore the `after_setup_theme` action must be used for this.

The problem with the existing code is not that there is no theme set up. The problem is that the _incorrect_ theme is set up at that point when considering the live preview.

Reply:
To my previous points, please also see https://developer.wordpress.org/reference/hooks/after_setup_theme/:
> [This action] is generally used to perform basic setup, registration, and init actions for a theme.

`add_theme_support()` is part of this basic setup, and should not be used on `setup_theme`. Themes are not even able to hook into `setup_theme` as it runs before the current theme files are being loaded.